lib/poller.js

```diff
@@ -163,14 +163,57 @@ class Poller {
     }
 
     const secretManifest = await this._createSecretManifest()
-    this._logger.info(`upserting secret ${this._namespace}/${this._name}`)
+    const kubeSecret = kubeNamespace.secrets(this._name)
+    let existingSecret
 
     try {
-      return await kubeNamespace.secrets.post({ body: secretManifest })
+      this._logger.info(`getting secret ${this._namespace}/${this._name}`)
+      const secretResponse = await kubeSecret.get()
+      existingSecret = secretResponse.body
     } catch (err) {
-      if (err.statusCode !== 409) throw err
-      return kubeNamespace.secrets(this._name).put({ body: secretManifest })
+      if (err.statusCode !== 404) throw err
+      // do nothing if the secret is not found
     }
+
+    if (existingSecret && this._equalSecretData(existingSecret, secretManifest)) {
+      this._logger.info(`skipping secret ${this._namespace}/${this._name} upsert, objects are the same`)
+      return Promise.resolve(true)
+    } else if (existingSecret === undefined) {
+      this._logger.info(`creating secret ${this._namespace}/${this._name}`)
+      return await kubeNamespace.secrets.post({ body: secretManifest })
+    } else {
+      this._logger.info(`updating secret ${this._namespace}/${this._name}`)
+      return kubeSecret.put({ body: secretManifest })
+    }
+  }
+
+  /**
+   * Checks if a secret and the desired secret manifest are equal
+   *
+   * @param   {Object}  kubeSecret      An actual kubernetes secret from the kube-client
+   * @param   {Object}  secretManifest  A hash representing the desired secret
+   *
+   * @return  {Boolean}                  Boolean if they are the same or not
+   */
+  _equalSecretData (kubeSecret, secretManifest) {
+    let result = true
+    const liveSecret = clonedeep(kubeSecret)
+    const desiredSecret = clonedeep(secretManifest)
+
+    // Only use annotations and labels for metadata checking
+    const secrets = [liveSecret, desiredSecret]
+    secrets.forEach((s) => {
+      s.metadata = {
+        labels: s.metadata.labels,
+        annotations: s.metadata.annotations
+      }
+    })
+
+    result = result ? JSON.stringify(secretManifest.metadata) === JSON.stringify(kubeSecret.metadata) : false
+
+    // check secret data
+    result = result ? JSON.stringify(kubeSecret.data) === JSON.stringify(secretManifest.data) : false
```

Comment:
here, too

Reply:
oh my, so sorry for missing it 😅 updated!